api/app/src/command/webhook/withings.ts

```diff
@@ -0,0 +1,87 @@
+import dayjs from "dayjs";
+import { ProviderSource } from "@metriport/api";
+import { getConnectedUserByTokenOrFail } from "../connected-user/get-connected-user";
+import { getSettingsOrFail } from "../settings/getSettings";
+import { Constants } from "../../shared/constants";
+import { ConnectedUser } from "../../models/connected-user";
+import { createWebhookRequest } from "./webhook-request";
+import { processRequest, WebhookUserDataPayload, reportUsage } from "./webhook";
+import { capture } from "../../shared/notifications";
+
+export type WithingsWebhook = {
+  userid: string;
+  date?: string;
+  startdate?: string;
+  enddate?: string;
+  appli: string;
+};
+
+const activityCategory = "16";
+const bodyCategory = "1";
+const biometricsCategories = ["2", "4", "54", "58"];
+const sleepCategory = "44";
+
+export const processData = async (data: WithingsWebhook) => {
+  const categoryNum = data.appli;
+  const withingsUserId = data.userid;
+  let startdate = "";
+
+  if (data.startdate) {
+    startdate = dayjs.unix(parseInt(data.startdate)).format();
+  } else if (data.date) {
+    startdate = data.date;
+  }
+
+  try {
+    const connectedUser = await getConnectedUserByTokenOrFail(
+      ProviderSource.withings,
+      withingsUserId
+    );
+
+    const cxId = connectedUser.cxId;
+    const cxUserId = connectedUser.cxUserId;
+
+    const settings = await getSettingsOrFail({ id: cxId });
+    const withingsData = await mapData(categoryNum, connectedUser, startdate);
+    const payload = { users: [{ userId: cxUserId, ...withingsData }] };
+    const webhookRequest = await createWebhookRequest({ cxId, payload });
+
+    await processRequest(webhookRequest, settings);
+    reportUsage(connectedUser.cxId, [connectedUser.cxUserId]);
+  } catch (error) {
+    capture.error(error, {
+      extra: { context: `webhook.withings.processData` },
+    });
+  }
+};
+
+export const mapData = async (
+  categoryNum: string,
+  connectedUser: ConnectedUser,
+  startdate: string
+): Promise<WebhookUserDataPayload> => {
+  const payload: WebhookUserDataPayload = {};
+  const provider = Constants.PROVIDER_MAP[ProviderSource.withings];
+
+  if (categoryNum === activityCategory) {
+    const activity = await provider.getActivityData(connectedUser, startdate);
+    payload.activity = [activity];
+  }
```

Comment:
ah so this problem is slightly different than Apple Health...

Apple will send us new info only, but here we're querying for the full day of data every time Withings tells us to.

How would the customer know what part of the payload is "new" data?

Reply:
So i tried to think of other ways around this but the webhook payload itself is structured as 'YYYY-MM-DD' with no time and the only way to query activity and sleep is by the entire day. I thought maybe i could create a timestamp of when we receive the payload and use that to filter the query for only new data but thats not very reliable because the webhook payload is not received after the data is inputed its pretty random.

Reply:
For workouts and sleep, most of the time it should only be one for a day so i could technically use the timestamp approach and when we receive the payload filter out only new ones in the last like 30 minutes

Reply:
Not sure if it helps, but aren't the readings/entries idempotent, as in: "these are the readings with IDs... you might get the same reading more than once, make sure to not duplicate them".

If so, that means we could read all the entries of the day everytime we get a notification from Withings... and our customers should be aware that they might get the same reading more than once.

Reply:
Thinking about this some more, I guess the customer could de-dup based on date right?

So, if they receive data for "today" twice, they in theory could discard all of previous data from "today". If this is reasonable, we just need to document accordingly to set expectations.

Reply:
Or, if we don't have IDs to identify those, could we document that webhook calls with Withings data should replace the previous data for the respective date?

Reply:
Just to clarify only Activity and Sleep return the whole day. Body and Biometrics will return only new data

Reply:
discussed offline under https://metriport.slack.com/archives/C04CXLSAF7V/p1683314443991239